api/kv_v2.go

```diff
@@ -551,3 +625,69 @@ func readThenWrite(ctx context.Context, client *Client, mountPath string, secret
 
 	return updatedSecret, nil
 }
+
+func setMetadataMap(metadata KVMetadataInput) (map[string]interface{}, error) {
+	md := make(map[string]interface{})
+	casRequiredKey := "cas_required"
+	deleteVersionAfterKey := "delete_version_after"
+	maxVersionsKey := "max_versions"
+	customMetadataKey := "custom_metadata"
+
+	// --[Explicit Zero Value For The User, Explicit Null On The Backend]--
+	// The KVMetadataInput struct is designed to have pointer fields so that
+	// the user can easily express the desire to explicitly set a field back to its zero
+	// value (e.g. false), as opposed to just having the field remain
+	// unchanged (e.g. nil).
+	// However, in the actual JSONMergePatch request performed by Logical,
+	// the meanings are reversed, where a null value means to delete the
+	// value at that field, and a zero value means to do nothing.
```

Comment:
Would it be confusing for users to have the semantics reversed for two different functions within the same library? Perhaps we should look at other PATCH methods as examples.

Reply:
The reason it behaves differently is because the Logical().JSONMergePatch method accepts a map, and this method accepts a struct. With the map can just choose to not include those values if you don't want to modify them, but with a struct, an unset value is `nil` or the zero value, which both mean something specific in JSON merge patch semantics. (Which is why I create my own map out of what I believe they intended with their KVMetadataInput struct, thanks to pointers)

The reason I chose for it to be a struct is just the standard "why statically typed things are nice" argument, in that users can easily see which metadata fields they even have the ability to modify (because as you can see, it's not _all_ of the metadata fields), and they don't have to worry about messing up by accidentally calling the map key `delete_versions_after` instead of `delete_version_after` like I did 500 times. 😆 

We could allow them to do a map instead like Logical().JSONMergePatch, but alas, we'd lose that built-in field knowledge I suppose. I guess the question is, which aspect of the UX is more important (knowing what fields you can edit, or not having to deal with pointers).

Your comment here is definitely worth thinking about though! Curious what others think here too!

(And yeah I'll look at some other PATCH methods, but my experiences with PATCH at previous companies always seemed to be situations where the teams had designed their code in a way where they passed around structs but didn't use pointers, and had to do crazy hacky things to get the zero value working, haha... so I think for PATCH it's a question of either maps or structs-with-pointers but not plain structs.)

Reply:
In general, a null value is interpreted as "remove this field" in JSON merge patches. That's not actually how Vault's patch handling works, however. Vault will not blindly remove all fields whose values are null as that would leave the resource in a bad state. For example, that would allow a user to remove the `created_at` field in a KVv2 metadata entry. Vault actually translates null into the zero value of that type. This happens in `FieldData.getPrimitive`, I believe. A field will only be removed for arbitrary maps, like underlying fields in `custom_metadata`.

Reply:
Thanks for the detailed explanations! This is indeed a tricky issue to get around...

Reply:
@ccapurso Gotcha! Okay thanks for the clarification there. So I think I can remove the "switch these fields to nil if they're the default value" bit here, since the null only comes into play with custom_metadata. I'll just change this to "if the field in the struct wasn't nil, then use the value (of the deref'd pointer), whatever it was". Thanks!